### ISSUE TITLE
Is there a simpler way to implement Circular Padding for Conv1d?

### ISSUE THREAD

**Issue by flatcapital** [huggingface/candle#849]
Hi, I'm enjoying the rapid progress on this library!

I'm implementing an embedding layer for a timeseries model ([TimesNet](https://github.com/thuml/Time-Series-Library/blob/main/tutorial/TimesNet_tutorial.ipynb)) and was wondering if there is an easier way to implement it compared to my approach below? ([PyTorch version](https://github.com/thuml/Time-Series-Library/blob/main/layers/Embed.py)) for reference.

What I'm wondering most about is the implementation for circular padding required for the conv1d input, as well as having a no bias conv1d layer. (I only saw no bias conv2d layer implemented). Also I think I would need to implement the Leaky ReLu since the  `DEFAULT_KAIMING_NORMAL` uses ReLu.

Essentially the following in Pytorch:
```
        self.tokenConv = nn.Conv1d(in_channels=c_in, out_channels=d_model,
                                   kernel_size=3, padding=padding, padding_mode='circular', bias=False)
        for m in self.modules():
            if isinstance(m, nn.Conv1d):
                nn.init.kaiming_normal_(
                    m.weight, mode='fan_in', nonlinearity='leaky_relu')

```

**My Implementation:**

```
use candle_core::{Device, DType, Module, Result, Tensor};
use candle_nn::{Conv1d, Conv1dConfig, Dropout, init, VarBuilder};

pub fn conv1d_no_bias(
    in_channels: usize,
    out_channels: usize,
    kernel_size: usize,
    cfg: Conv1dConfig,
    vb: crate::VarBuilder,
) -> Result<Conv1d> {
    let init_ws = init:: DEFAULT_KAIMING_NORMAL;
    let ws = vb.get_with_hints(
        (out_channels, in_channels / cfg.groups, kernel_size),
        "weight",
        init_ws,
    )?;
    Ok(Conv1d::new(ws, None, cfg))
}

fn pad_circular(input: &Tensor, padding: Vec<usize>) -> Result<Tensor> {
    let input_ndim = input.rank();
    let ndim_padded = padding.len() / 2;
    let ndim_nonpadded = input_ndim - ndim_padded;

    assert!(ndim_nonpadded == 1 || ndim_nonpadded == 2, "Invalid padding size.");

    let mut out_shape = input.dims().to_vec();

    // Calculate out shape
    for i in (0..ndim_padded).rev() {
        let pad_l = padding[2 * i];
        let pad_r = padding[2 * i + 1];
        let size = input.dim(ndim_nonpadded + i)?;
        out_shape[ndim_nonpadded + i] = size + pad_l + pad_r;

        assert!(pad_l <= size && pad_r <= size, "Padding value causes wrapping around more than once.");
    }

    let mut out = Tensor::zeros(out_shape.clone(), DType::F32, &Device::Cpu)?;
    let mut sections: Vec<Tensor> = Vec::new();

    for i in (0..ndim_padded).rev() {
        let dim = ndim_padded - i + ndim_nonpadded - 1;
        let pad_l = padding[2 * i];
        let pad_r = padding[2 * i + 1];

        sections.clear();

        // Left padding
        if pad_l > 0 {
            let src = input.narrow(dim, input.dim(dim)? - pad_l, pad_l)?;
            sections.push(src);
        }

        sections.push(input.clone());

        // Right padding
        if pad_r > 0 {
            let src = input.narrow(dim, 0, pad_r)?;
            sections.push(src);
        }

        out = Tensor::cat(&sections.iter().collect::<Vec<_>>(), dim)?;
    }

    Ok(out)
}

fn sinusoids(dim: usize, max_len: usize) -> Result<Tensor> {
    let max_timescale = 10000f32;
    let log_timescale_increment = -(max_timescale.ln() / dim as f32);

    let div_term: Vec<f32> = (0..dim)
        .step_by(2)
        .map(|i| (i as f32 * log_timescale_increment).exp())
        .collect();

    let mut interleaved_values: Vec<f32> = vec![0.0; max_len * dim];
    
    for i in 0..max_len {
        for j in 0..(dim / 2) {
            let scaled_time = i as f32 * div_term[j];
            interleaved_values[i * dim + j * 2] = scaled_time.sin();
            interleaved_values[i * dim + j * 2 + 1] = scaled_time.cos();
        }
    }

    let interleaved = Tensor::from_vec(interleaved_values, vec![max_len, dim], &Device::Cpu)?.unsqueeze(0)?;
    Ok(interleaved)
}

#[derive(Debug)]
pub struct DataEmbedding {
    value_embed: Conv1d,
    pos_embed: Tensor,
    dropout: Dropout
}

impl DataEmbedding {

    pub fn new(vb: VarBuilder, in_channels: usize, dim: usize, dropout: f32) -> Result<Self> {

        let config = Conv1dConfig {
            padding: 0,
            ..Default::default()
        };
        let value_embed = conv1d_no_bias(in_channels, dim, 3, config, vb.pp("value_embed"))?;
        let pos_embed = sinusoids(dim, 5000)?;
        let dropout = Dropout::new(dropout);

        Ok(Self {
            value_embed,
            pos_embed,
            dropout
        })
    }
}

impl Module for DataEmbedding {
    fn forward(&self, x: &Tensor) -> Result<Tensor> {
        let x = x.permute([0, 2, 1])?;  // [B, C, T] for conv1d
        let x = pad_circular(&x, vec![1,1])?;
        let x = self.value_embed.forward(&x)?;
        let pos_embed = self.pos_embed.narrow(1, 0, x.dims()[2])?.permute([0, 2, 1])?;
        let x = x.broadcast_add(&pos_embed)?;
        let x = x.permute([0, 2, 1])?;  // [B, T, C]
        let x = self.dropout.forward(&x, true)?;
        Ok(x)
    }
}
```

**Comment by LaurentMazare** [comment]
This looks like a pretty good approach, I feel that most of the comments I would have are more about being rust idiomatic than candle.
Fwiw here is a fairly short implementation, very similar to what you did but that only supports the necessary bits.
```rust
fn pad_circular1d(xs: &Tensor, pad_left: usize, pad_right: usize) -> Result<usize> {
    let (b_size, channels, len) = xs.dims3()?;
    if len < pad_left || len < pad_right {
        candle::bail!("tensor is too small {len} for padding {pad_left}, {pad_right}")
    }
    let left = xs.narrow(2, len - pad_right, pad_right)?;
    let right = xs.narrow(2, 0, pad_left)?;
    Tensor::cat(&[&left, xs, &right], 2)
}
```

**Comment by flatcapital** [comment]
That is great, thanks for the more concise solution. Much appreciated!
